server/src/main/java/org/opensearch/search/approximate/ApproximateScoreQuery.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.approximate;
+
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Weight;
+import org.opensearch.search.internal.SearchContext;
+
+import java.io.IOException;
+
+/**
+ * Entry-point for the approximation framework.
+ * This class is heavily inspired by {@link org.apache.lucene.search.IndexOrDocValuesQuery}. It acts as a wrapper that consumer two queries, a regular query and an approximate version of the same. By default, it executes the regular query and returns {@link Weight#scorer} for the original query. At run-time, depending on certain constraints, we can re-write the {@code Weight} to use the approximate weight instead.
+ */
+public class ApproximateScoreQuery extends Query {
+
+    private final Query originalQuery;
+    private final ApproximateQuery approximationQuery;
+
+    protected Query resolvedQuery;
+
+    public ApproximateScoreQuery(Query originalQuery, ApproximateQuery approximationQuery) {
+        this.originalQuery = originalQuery;
+        this.approximationQuery = approximationQuery;
+    }
+
+    public Query getOriginalQuery() {
+        return originalQuery;
+    }
+
+    public ApproximateQuery getApproximationQuery() {
+        return approximationQuery;
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
+        if (resolvedQuery == null) {
+            throw new IllegalStateException("Cannot create weight without setContext being called");
+        }
+        return resolvedQuery.createWeight(searcher, scoreMode, boost);
+    }
```

Comment:
So, we shouldn't really implement `createWeight` here. Instead, we can just throw an exception (which is what `Query` does anyway -- take a look at `MultiTermQuery`, which also doesn't implement `createWeight`).

just like `MultiTermQuery`, we should work by implementing `rewrite(IndexSearcher)`. The `rewrite` should return `resolvedQuery` (and throw an exception if it's null).

The basic idea is that rewriting happens before caching, and the rewritten query is used as the cache key. If you get the original query (which is cacheable), you want the original query to be the key. If you get the approximate query (which maybe could be cacheable), you want that to be the key. 

You absolutely never want to return an approximate cached value for an exact query, which could happen if `ApproximateScoreQuery` is the key. By implementing `rewrite`, we make sure that it never is.

Reply:
Got it! I'm essentially calling `resolvedQuery.rewrite(indexSearcher)` to return the `resolvedQuery`